diff --git a/src/HttpServerBodyCommand.cc b/src/HttpServerBodyCommand.cc
index e9e7b92a..1ef48f7f 100644
--- a/src/HttpServerBodyCommand.cc
+++ b/src/HttpServerBodyCommand.cc
@@ -149,6 +149,7 @@ void HttpServerBodyCommand::addHttpServerResponseCommand(bool delayed)
   auto resp = make_unique<HttpServerResponseCommand>(getCuid(), httpServer_, e_,
                                                      socket_);
   if (delayed) {
+    e_->deleteSocketForWriteCheck(socket_, resp.get());
     e_->addCommand(
         make_unique<DelayedCommand>(getCuid(), e_, 1_s, std::move(resp), true));
     return;
