diff --git a/src/core/or/or.h b/src/core/or/or.h
index 088c45342b..265b84425a 100644
--- a/src/core/or/or.h
+++ b/src/core/or/or.h
@@ -899,7 +899,7 @@ typedef struct or_state_t or_state_t;
 /********************************* circuitbuild.c **********************/
 
 /** How many hops does a general-purpose circuit have by default? */
-#define DEFAULT_ROUTE_LEN 3
+#define DEFAULT_ROUTE_LEN 2
 
 /* Circuit Build Timeout "public" structures. */
 
